com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerModelLoader.cs

```diff
@@ -23,53 +27,92 @@ public static class ControllerModelLoader
+        /// <returns>A gameobject representing the generated controller model in the scene</returns>
+        public async static Task<GameObject> TryGenerateControllerModelFromPlatformSDK(InputDevice inputDevice, Handedness handedness)
         {
+            // Sanity check to ensure that the xrInputDevice's usages matches the provided handedness
+            InternedString targetUsage = handedness == Handedness.Left ? CommonUsages.LeftHand : CommonUsages.RightHand;
+            Debug.Assert(inputDevice.usages.Contains(targetUsage));
+
+            // Proceed with trying to load the model from the platform
             GameObject gltfGameObject = null;
 
 #if MROPENXR_PRESENT && (UNITY_STANDALONE_WIN || UNITY_WSA || UNITY_ANDROID) && GLTFAST_PRESENT && KTX_PRESENT
             ControllerModel controllerModelProvider = handedness == Handedness.Left ? ControllerModel.Left :
                                                         handedness == Handedness.Right ?  ControllerModel.Right :
                                                         null;
-
             if (controllerModelProvider == null)
             {
                 Debug.LogWarning("Controller model provider does not exist for this handedness");
                 return null;
             }
 
+            // Try to obtain a model key
             if (!controllerModelProvider.TryGetControllerModelKey(out ulong modelKey))
             {
-                Debug.LogWarning("Failed to obtain controller model key from platform.");
+                if(!warningCache.ContainsKey(inputDevice) || !warningCache[inputDevice])
+                {
+                    Debug.LogWarning("Failed to obtain controller model key from platform.");
+                }
+
+                // Make sure we record in our cache that we've previously raised a warning for this input device
+                warningCache[inputDevice] = true;
+
                 return null;
             }
 
-            if (controllerModelDictionary.TryGetValue(modelKey, out gltfGameObject))
+            // Make sure we record in our cache that we've succeeded in loading this model without warnings
+            warningCache[inputDevice] = false;
+
+            // Check if a gameobject already exists for this modelkey. If so, set it active and return it
+            if (controllerModelDictionary.TryGetValue(modelKey, out gltfGameObject) && gltfGameObject != null)
             {
                 gltfGameObject.SetActive(true);
                 return gltfGameObject;
             }
 
+            // Otherwise, try to load the model from the model provider
             byte[] modelStream = await controllerModelProvider.TryGetControllerModel(modelKey);
-
             if (modelStream == null || modelStream.Length == 0)
             {
-                Debug.LogError("Failed to obtain controller model from platform.");
+                if (!errorCache.ContainsKey(modelKey) || !errorCache[modelKey])
```

Comment:
You could also do `!errorCache.TryGetValue(modelKey, out bool errorLogged) || !errorLogged` here to reduce the lookups to one.

Reply:
Done